### ISSUE TITLE
ConsumerImpl.redeliverUnacknowledgedMessages() get empty messageids

### ISSUE THREAD

**Issue by jiazhai** [apache/pulsar#3287]
#### Expected behavior

no panic and error.

#### Actual behavior
meet error info:
```
java.util.NoSuchElementException: No value present
        at java.util.Optional.get(Optional.java:135)
        at org.apache.pulsar.client.impl.ConsumerImpl.redeliverUnacknowledgedMessages(ConsumerImpl.java:1246)
        at org.apache.pulsar.client.impl.UnAckedMessageTracker$1.run(UnAckedMessageTracker.java:96)
        at org.apache.pulsar.shade.io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663)
        at org.apache.pulsar.shade.io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738)
        at org.apache.pulsar.shade.io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466)
        at org.apache.pulsar.shade.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
```

And some times, there is log: `0 messages have timed-out`.

#### Steps to reproduce

How can we reproduce the issue

#### System configuration
**Pulsar version**: 2.2.1

**Comment by jiazhai** [comment]
In 2.2.1 the redeliver is triggered as this code:
```
    public void start(PulsarClientImpl client, ConsumerBase<?> consumerBase, long ackTimeoutMillis) {
        this.stop();
        timeout = client.timer().newTimeout(new TimerTask() {
            @Override
            public void run(Timeout t) throws Exception {
                if (isAckTimeout()) {
                    log.warn("[{}] {} messages have timed-out", consumerBase, oldOpenSet.size());
                    Set<MessageId> messageIds = new HashSet<>();
                    oldOpenSet.forEach(messageIds::add);
                    oldOpenSet.clear();
                    consumerBase.redeliverUnacknowledgedMessages(messageIds);   < ===
                }
                toggle();
                timeout = client.timer().newTimeout(this, ackTimeoutMillis, TimeUnit.MILLISECONDS);
            }
        }, ackTimeoutMillis, TimeUnit.MILLISECONDS);
    }

```
As the code, messageIds could be empty. and caused the issue.

While in latest 2.3.0, @codelipenghui added some change in PR #3118, and there is a check of condition `if (messageIds.size() > 0)` now.

```
   public UnAckedMessageTracker(PulsarClientImpl client, ConsumerBase<?> consumerBase, long ackTimeoutMillis, long tickDurationInMs) {
        Preconditions.checkArgument(tickDurationInMs > 0 && ackTimeoutMillis >= tickDurationInMs);
        this.ackTimeoutMillis = ackTimeoutMillis;
        this.tickDurationInMs = tickDurationInMs;
        ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
        this.readLock = readWriteLock.readLock();
        this.writeLock = readWriteLock.writeLock();
        this.messageIdPartitionMap = new ConcurrentHashMap<>();
        this.timePartitions = new LinkedList<>();

        int blankPartitions = (int)Math.ceil((double)this.ackTimeoutMillis / this.tickDurationInMs);
        for (int i = 0; i < blankPartitions + 1; i++) {
            timePartitions.add(new ConcurrentOpenHashSet<>());
        }

        timeout = client.timer().newTimeout(new TimerTask() {
            @Override
            public void run(Timeout t) throws Exception {
                Set<MessageId> messageIds = new HashSet<>();
                writeLock.lock();
                try {
                    timePartitions.addLast(new ConcurrentOpenHashSet<>());
                    ConcurrentOpenHashSet<MessageId> headPartition = timePartitions.removeFirst();
                    if (!headPartition.isEmpty()) {
                        log.warn("[{}] {} messages have timed-out", consumerBase, timePartitions.size());
                        headPartition.forEach(messageId -> {
                            messageIds.add(messageId);
                            messageIdPartitionMap.remove(messageId);
                        });
                    }
                } finally {
                    writeLock.unlock();
                }
                if (messageIds.size() > 0) {   < ===
                    consumerBase.redeliverUnacknowledgedMessages(messageIds);
                }
                timeout = client.timer().newTimeout(this, tickDurationInMs, TimeUnit.MILLISECONDS);
            }
        }, this.tickDurationInMs, TimeUnit.MILLISECONDS);
    }
```

**Comment by jiazhai** [comment]
@sijie #3118 is another change, and it is not intention to do the fix for this issue. If we want to do the fix in 2.2 or earlier, we may need another PR.

And in the 2.2.1 code:
```
    public void start(PulsarClientImpl client, ConsumerBase<?> consumerBase, long ackTimeoutMillis) {
        this.stop();
        timeout = client.timer().newTimeout(new TimerTask() {
            @Override
            public void run(Timeout t) throws Exception {
                if (isAckTimeout()) {          < === 1
                    log.warn("[{}] {} messages have timed-out", consumerBase, oldOpenSet.size());  < === 2
                    Set<MessageId> messageIds = new HashSet<>();
                    oldOpenSet.forEach(messageIds::add);
                    oldOpenSet.clear();
                    consumerBase.redeliverUnacknowledgedMessages(messageIds);   < ===
                }
                toggle();
                timeout = client.timer().newTimeout(this, ackTimeoutMillis, TimeUnit.MILLISECONDS);
            }
        }, ackTimeoutMillis, TimeUnit.MILLISECONDS);
    }
``` 
In [isAckTimeout](https://github.com/apache/pulsar/blob/v2.2.1-candidate-2/pulsar-client/src/main/java/org/apache/pulsar/client/impl/UnAckedMessageTracker.java#L162), it has already checked `oldOpenSet.isEmpty()`,  so the race window between "< ===1" and "< ===2" should be really small. it should be hard to trigger.

**Comment by jiazhai** [comment]
checked with the reporter again. They are using 2.2.1 server, but with 2.2.0 client. And this issue is caused by a wrongly change of above `toggle` orer, and it should be fixed in 2.2.1. 
will close this issue.
